### DISCUSSION TITLE
TRUSTED_PROXIES not working in LAMP stack

### DISCUSSION THREAD

**Discussion by kr-nn** [firefly-iii/firefly-iii#7582] [Firefly III] [answered]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely a Firefly III issue, not me](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

My topology is:
HAPROXY (OPNsense) > Apache > Firefly (LAMP, no docker) 

I have tried both TRUSTED_PROXIES=* and **
It seems like everything I do I get the following errors:
`fused to set the document's base URI to 'http://money.redacted.xyz/' because it violates the following Content Security Policy directive: "base-uri 'self'".`<br>
`refused to send form data to 'http://money.redacted.xyz/login' because it violates the following Content Security Policy directive: "form-action 'self'".`

I don't know why Firefly is requesting this over http when my URL is https:
![image](https://github.com/firefly-iii/firefly-iii/assets/84939608/3750ecac-fbb6-4ff3-9f63-175934b62aef)

A number of files are requested over http and not https. My browser forces https in these instances.
If I disable CSP from the below .env file I can login but my browser also blocks http requests.
If there's a toggle to force https requests that would be helpful, but from looking for a couple weeks it seems that's not a thing, and trusted_proxies is the option I'm supposed to use.

Note: I'm using LXC (Proxmox) So I'd like to avoid docker as much as I can.

# HAPROXY:
```
frontend Secure_Endpoint
    bind 0.0.0.0:443 name 0.0.0.0:443 ssl  crt-list /tmp/haproxy/ssl/64309303e25b97.03965681.certlist 
    mode http
    option http-keep-alive
    default_backend Silent-Drop
    option forwardfor

    # ACL: Firefly FQDN
    acl acl_6473d9a9d1f630.02505619 hdr_beg(host) -i money.redacted.xyz
    # ACTION: Firefly-E
    use_backend Firefly-E if acl_6473d9a9d1f630.02505619

frontend Insecure_Endpoint
    bind 0.0.0.0:80 name 0.0.0.0:80 
    mode http
    option http-keep-alive
    default_backend Silent-Drop

    # ACL: Firefly FQDN
    acl acl_6473d9a9d1f630.02505619 hdr_beg(host) -i money.redacted.xyz
    # ACTION: Firefly-U
    use_backend Firefly-U if acl_6473d9a9d1f630.02505619

# Backend: Firefly-U (https://10.0.10.43)
backend Firefly-U
    # health checking is DISABLED
    mode http
    balance source
    http-reuse safe
    server Firefly-U 10.0.10.43:80 

# Backend: Firefly-E (https://10.0.10.43)
backend Firefly-E
    # health checking is DISABLED
    mode http
    balance source
    http-reuse safe
    server Firefly-E 10.0.10.43:443 
```

# APACHE:
```
<VirtualHost *:80>
        DocumentRoot /var/www/html
        ErrorLog ${APACHE_LOG_DIR}/error.log
        CustomLog ${APACHE_LOG_DIR}/access.log combined
</VirtualHost>
<VirtualHost *:443>
        DocumentRoot /var/www/html
        CustomLog ${APACHE_LOG_DIR}/access.log combined
</VirtualHost>
```

# FIREFLY:
```
# Redacted to just the relevant environment variables

TRUSTED_PROXIES=**
#DISABLE_CSP_HEADER=true
```

### Debug information

Debug information generated at 2023-06-02 00:00:00 America/Toronto for Firefly III version **6.0.5**.

| Scope | Version |
| --- | --- |
| Firefly III | 6.0.5 |
| Firefly III API | 2.0.1 |
| Build | (unknown), (unknown) |
| DB version | 19 (exp. 19) |
| PHP | `8.2.6` |
| Host | `Linux` |

| System info | Value |
| --- | --- |
| System TZ | America/Toronto |
| Browser TZ | America/Toronto |
| App environment | local |
| App debug mode | false |
| App cache driver | file |
| App logging | notice, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Max upload | 2097152 (2 MB) |
| Interface | apache2handler |
| Default language | en_US |
| Default locale | equal |
| BCscale | 12 |
| DB drivers | mysql |
| Current driver | mysql |
| Trusted proxies (.env) | `**` |

| User info | Value |
| --- | --- |
| Login provider | eloquent  |
| User guard | web  |
| Headers | N/A, N/A  |
| Stateful domains |  |
| Session start |  |
| Session end |  |
| Session first |  |
| User ID | 1 |
| User language | en_US |
| User locale | en_US |
| Attempt at "en_US.utf8" | 'en_US.utf8' |
| Attempt at "en_US.UTF-8" | 'en_US.UTF-8' |
| User agent | Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/112.0.0.0 Safari/537.36 |

### Expected behaviour

1. Http Requests to be changed to https
2. CSP to work correctly

**Comment by JC5** [comment, marked as answer]
Ultimately, this is a configuration error, not a bug. The equivalent nginx configuration is as follows. If these headers are correct from the proxy to Apache, it should work:

```
location / {
    proxy_set_header X-Forwarded-Host $host;
    proxy_set_header X-Forwarded-Server $host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header Host $host;
}
```

**Reply by kr-nn** [reply]
Thanks this resolved my issue.

For everyone else who comes across this.
X-Forwarded-Proto was what was needed.

If you're getting CSP errors in console and the URI says HTTP it's likely that header wasn't working.
Check and make sure all your headers are working:
`tcpdump -i eth0 -s 0 -A 'tcp dst port 80 or tcp dst port 443' > /dump.txt`

In my case I had to explicitly say https instead of `%[req.hdr(scheme)]`

@JC5 Thanks again

**Reply by Dinth** [reply]
I have tried that, unfortunately it doesnt work.
When i add all 5 proxy_set_header im getting 502 bad request error.
If i omit 
`    proxy_set_header Host $host;`
leaving the other 4 lines im getting a weird glitch while logging in, where Firefox says that the connection is not secure underneath of the password field, but also says that the connection is secure in the address bar. 
Any advice please?
Im using slightly modified Firefly III config from linuxserver/swag:

![image](https://github.com/firefly-iii/firefly-iii/assets/313674/393289ba-1058-4a61-a807-7befbb910edc)

**Reply by kr-nn** [reply]
this is the headers I set

http-request set-header X-Forwarded-Proto https 
http-request set-header X-Forwarded-Server %[req.hdr(host)] 
http-request set-header X-Forwarded-Host %[req.hdr(host)] 
http-request set-header Host %[req.hdr(host)] 
http-request set-header X-Forwarded-For %[src] 

note where I said:

> In my case I had to explicitly say https instead of %[req.hdr(scheme)]
